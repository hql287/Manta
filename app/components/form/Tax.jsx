// Libraries
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { isEqual } from 'lodash';
// Custom Components
import { Section, Header } from '../shared/Section';
// Animation
import _withFadeInAnimation from '../shared/hoc/_withFadeInAnimation';
// Translation
import * as TRANSLATION_LABELS from '../../constants/translations';
// Styles
import styled from 'styled-components';
const TaxID = styled.div``;
const TaxAmount = styled.div`
  flex: 1;
`;

const Form = styled.div`
  padding: 20px;
  background: #f9fafa;
  border-radius: 4px;
  margin-bottom: 20px;
  border: 1px solid #f2f3f4;
`;

const Row = styled.div`
  display: flex;
  margin: 0 -15px;
`;

const Field = styled.div`
  flex: 1;
  margin: 0 15px 20px 15px;
`;

// Component
export class Tax extends Component {
  constructor(props) {
    super(props);
    this.state = props.tax;
    this.isSettingsSaved = this.isSettingsSaved.bind(this);
    this.saveAsDefault = this.saveAsDefault.bind(this);
    this.handleInputChange = this.handleInputChange.bind(this);
  }

  // Handle Form Clear
  componentWillReceiveProps(nextProps) {
    // Already made changes but not saved
    if (this.state !== this.props.savedSettings) {
      // Reset to savedSettings if the below confition is met
      if (nextProps.tax === nextProps.savedSettings) {
        this.setState(nextProps.savedSettings, () => {
          this.updateTaxState();
        });
      }
    }
  }

  shouldComponentUpdate(nextProps, nextState) {
    if (this.state !== nextState) return true;
    if (this.props.tax !== nextProps.tax) return true;
    if (this.props.savedSettings !== nextProps.savedSettings) return true;
    return false;
  }

  handleInputChange(event) {
    const target = event.target;
    const name = target.name;
    const value = name === 'amount' ? parseFloat(target.value) : target.value;
    this.setState(
      {
        [name]: value,
      },
      () => {
        this.updateTaxState();
      }
    );
  }

  updateTaxState() {
    const { updateFieldData } = this.props;
    updateFieldData('tax', this.state);
  }

  isSettingsSaved() {
    return isEqual(this.state, this.props.savedSettings);
  }

  saveAsDefault() {
    const { updateSavedSettings } = this.props;
    updateSavedSettings('tax', this.state);
  }

  render() {
    return (
      <Section>
        <Header>
          <label className="itemLabel">{ this.props.translate(TRANSLATION_LABELS.TAXFRM_HEADING) }</label>
          {!this.isSettingsSaved() && (
            <a href="#" onClick={this.saveAsDefault}>
              <i className="ion-checkmark" /> { this.props.translate(TRANSLATION_LABELS.TAXFRM_SAVEDEFAULT) }
            </a>
          )}
        </Header>
        <Form>
          <Row>
            <Field>
              <label className="itemLabel">{ this.props.translate(TRANSLATION_LABELS.TAXFRM_TAX_ID) }</label>
              <TaxID>
                <input
                  name="tin"
                  type="text"
                  value={this.state.tin}
                  onChange={this.handleInputChange}
                  placeholder={ this.props.translate(TRANSLATION_LABELS.TAXFRM_TAX_ID_PLACEHOLDER) }
                />
              </TaxID>
            </Field>
          </Row>
          <Row>
            <Field>
              <label className="itemLabel">{ this.props.translate(TRANSLATION_LABELS.TAXFRM_AMOUNT_PCT) }</label>
              <TaxAmount>
                <input
                  name="amount"
                  type="number"
                  step="0.01"
                  value={this.state.amount}
                  onChange={this.handleInputChange}
                  placeholder={ this.props.translate(TRANSLATION_LABELS.TAXFRM_AMOUNT_PCT_PLACEHOLDER) }
                />
              </TaxAmount>
            </Field>
            <Field>
              <label className="itemLabel">{ this.props.translate(TRANSLATION_LABELS.TAXFRM_TAX_METHOD) }</label>
              <select
                name="method"
                value={this.state.method}
                onChange={this.handleInputChange}
              >
                <option value="default">{ this.props.translate(TRANSLATION_LABELS.TAXFRM_TAX_METHOD_DEFEAULT) }</option>
                <option value="reverse">{ this.props.translate(TRANSLATION_LABELS.TAXFRM_TAX_METHOD_REVERSE) }</option>
              </select>
            </Field>
          </Row>
        </Form>
      </Section>
    );
  }
}

Tax.propTypes = {
  updateSavedSettings: PropTypes.func.isRequired,
  tax: PropTypes.object.isRequired,
  savedSettings: PropTypes.object.isRequired,
  updateFieldData: PropTypes.func.isRequired,
  translate: PropTypes.func.isRequired,
  currentLanguage: PropTypes.string,
};

// Exports
export default _withFadeInAnimation(Tax);
