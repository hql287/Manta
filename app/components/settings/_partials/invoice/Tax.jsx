import React, { Component } from 'react';
import PropTypes from 'prop-types';

import { Row, Field, Part } from '../../../shared/Part';

function Tax({ t, tax, handleTaxChange }) {
  return [
    <label key="label" className="itemLabel">
      {t('settings:fields:taxSettings')}
    </label>,
    <Part key="part">
      <Row>
        <Field>
          <label className="itemLabel">{t('form:fields:tax:id')}</label>
          <input
            name="tin"
            type="text"
            value={tax.tin}
            onChange={handleTaxChange}
            placeholder={t('form:fields:tax:id')}
          />
        </Field>
      </Row>
      <Row>
        <Field>
          <label className="itemLabel">{t('common:amount')}</label>
          <input
            name="amount"
            type="number"
            step="0.01"
            value={tax.amount}
            onChange={handleTaxChange}
            placeholder={t('common:amount')}
          />
        </Field>
        <Field>
          <label className="itemLabel">{t('form:fields:tax:method')}</label>
          <select name="method" value={tax.method} onChange={handleTaxChange}>
            <option value="default">{t('common:default')}</option>
            <option value="reverse">{t('form:fields:tax:reverse')}</option>
          </select>
        </Field>
      </Row>
    </Part>,
  ];
}

Tax.propTypes = {
  tax: PropTypes.object.isRequired,
  handleTaxChange: PropTypes.func.isRequired,
  t: PropTypes.func.isRequired,
};

export default Tax;
