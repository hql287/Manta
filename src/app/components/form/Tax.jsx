// Libraries
import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import { isEqual } from 'lodash';
// Custom Components
import { Section, Header } from '../shared/Section';
// Animation
import _withFadeInAnimation from '../shared/hoc/_withFadeInAnimation';
// Styles
import styled from 'styled-components';
import { Part, Row, Field } from '../shared/Part';
const TaxID = styled.div``;
const TaxAmount = styled.div`flex: 1;`;

// Component
export class Tax extends PureComponent {
  constructor(props) {
    super(props);
    this.state = props.tax;
    this.isSettingsSaved = this.isSettingsSaved.bind(this);
    this.saveAsDefault = this.saveAsDefault.bind(this);
    this.handleInputChange = this.handleInputChange.bind(this);
  }

  // Handle Form Clear
  componentWillReceiveProps(nextProps) {
    // Already made changes but not saved
    if (this.state !== this.props.savedSettings) {
      // Reset to savedSettings if the below confition is met
      if (nextProps.tax === nextProps.savedSettings) {
        this.setState(nextProps.savedSettings, () => {
          this.updateTaxState();
        });
      }
    }
  }

  handleInputChange(event) {
    const target = event.target;
    const name = target.name;
    const value = name === 'amount' ? parseFloat(target.value) : target.value;
    this.setState(
      {
        [name]: value,
      },
      () => {
        this.updateTaxState();
      }
    );
  }

  updateTaxState() {
    const { updateFieldData } = this.props;
    updateFieldData('tax', this.state);
  }

  isSettingsSaved() {
    return isEqual(this.state, this.props.savedSettings);
  }

  saveAsDefault() {
    const { updateSavedSettings } = this.props;
    updateSavedSettings('tax', this.state);
  }

  render() {
    const { t } = this.props;
    return (
      <Section>
        <Header>
          <label className="itemLabel">{t('form:fields:tax:name')}</label>
          {!this.isSettingsSaved() && (
            <a href="#" onClick={this.saveAsDefault}>
              <i className="ion-checkmark" /> {t('common:saveAsDefault')}
            </a>
          )}
        </Header>
        <Part>
          <Row>
            <Field>
              <label className="itemLabel">{t('form:fields:tax:id')}</label>
              <TaxID>
                <input
                  name="tin"
                  type="text"
                  value={this.state.tin}
                  onChange={this.handleInputChange}
                  placeholder={t('form:fields:tax:id')}
                />
              </TaxID>
            </Field>
          </Row>
          <Row>
            <Field>
              <label className="itemLabel">{t('common:amount')} (%)</label>
              <TaxAmount>
                <input
                  name="amount"
                  type="number"
                  step="0.01"
                  value={this.state.amount}
                  onChange={this.handleInputChange}
                  placeholder={t('common:amount')}
                />
              </TaxAmount>
            </Field>
            <Field>
              <label className="itemLabel">{t('form:fields:tax:method')}</label>
              <select
                name="method"
                value={this.state.method}
                onChange={this.handleInputChange}
              >
                <option value="default">{t('common:default')}</option>
                <option value="reverse">{t('form:fields:tax:reverse')}</option>
              </select>
            </Field>
          </Row>
        </Part>
      </Section>
    );
  }
}

Tax.propTypes = {
  updateSavedSettings: PropTypes.func.isRequired,
  t: PropTypes.func.isRequired,
  tax: PropTypes.object.isRequired,
  savedSettings: PropTypes.object.isRequired,
  updateFieldData: PropTypes.func.isRequired,
};

// Exports
export default _withFadeInAnimation(Tax);
