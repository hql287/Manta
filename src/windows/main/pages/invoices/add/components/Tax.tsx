import React, { useState, useEffect } from 'react'
import { isEqual } from 'lodash'
import styled from 'styled-components'
import { Section, Header } from '@uiSharedComponents/Section'
import { Part, Row, Field } from '@uiSharedComponents/Part'
import _withFadeInAnimation from '@uiSharedComponents/hoc/_withFadeInAnimation'

const TaxID = styled.div``
const TaxAmount = styled.div`
  flex: 1;
`

interface TaxProps {
  t: (key: string) => string
  tax: { tin: string; amount: number; method: string }
  savedSettings: { tin: string; amount: number; method: string }
  updateFieldData: (field: string, value: object) => void
  updateSavedSettings: (field: string, value: object) => void
}

const Tax: React.FC<TaxProps> = ({
  t,
  tax,
  savedSettings,
  updateFieldData,
  updateSavedSettings,
}) => {
  const [state, setState] = useState(tax)

  useEffect(() => {
    if (!isEqual(state, savedSettings) && isEqual(tax, savedSettings)) {
      setState(savedSettings)
      updateFieldData('tax', savedSettings)
    }
  }, [tax, savedSettings, state, updateFieldData])

  const handleInputChange = (
    event: React.ChangeEvent<HTMLInputElement | HTMLSelectElement>
  ) => {
    const { name, value } = event.target
    const parsedValue = name === 'amount' ? parseFloat(value) : value

    setState((prevState) => {
      const newState = { ...prevState, [name]: parsedValue }
      updateFieldData('tax', newState)
      return newState
    })
  }

  const isSettingsSaved = () => isEqual(state, savedSettings)

  const saveAsDefault = () => {
    updateSavedSettings('tax', state)
  }

  return (
    <Section>
      <Header>
        <label className="itemLabel">{t('form:fields:tax:name')}</label>
        {!isSettingsSaved() && (
          <a href="#" onClick={saveAsDefault}>
            <i className="ion-checkmark" /> {t('common:saveAsDefault')}
          </a>
        )}
      </Header>
      <Part>
        <Row>
          <Field>
            <label className="itemLabel">{t('form:fields:tax:id')}</label>
            <TaxID>
              <input
                name="tin"
                type="text"
                value={state.tin}
                onChange={handleInputChange}
                placeholder={t('form:fields:tax:id')}
              />
            </TaxID>
          </Field>
        </Row>
        <Row>
          <Field>
            <label className="itemLabel">{t('common:amount')} (%)</label>
            <TaxAmount>
              <input
                name="amount"
                type="number"
                step="0.01"
                value={state.amount}
                onChange={handleInputChange}
                placeholder={t('common:amount')}
              />
            </TaxAmount>
          </Field>
          <Field>
            <label className="itemLabel">{t('form:fields:tax:method')}</label>
            <select
              name="method"
              value={state.method}
              onChange={handleInputChange}
            >
              <option value="default">{t('common:default')}</option>
              <option value="reverse">{t('form:fields:tax:reverse')}</option>
            </select>
          </Field>
        </Row>
      </Part>
    </Section>
  )
}

export default _withFadeInAnimation(Tax)
