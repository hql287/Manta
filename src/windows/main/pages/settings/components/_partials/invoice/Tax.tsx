import React from 'react'
import { Row, Field, Part } from '@uiSharedComponents/Part'

interface TaxProps {
  t: (key: string) => string
  tax: { tin: string; amount: number; method: string }
  handleTaxChange: (
    event: React.ChangeEvent<HTMLInputElement | HTMLSelectElement>
  ) => void
}

const Tax: React.FC<TaxProps> = ({ t, tax, handleTaxChange }) => {
  return (
    <>
      <label className="itemLabel">{t('settings:fields:taxSettings')}</label>
      <Part>
        <Row>
          <Field>
            <label className="itemLabel">{t('form:fields:tax:id')}</label>
            <input
              name="tin"
              type="text"
              value={tax?.tin}
              onChange={handleTaxChange}
              placeholder={t('form:fields:tax:id')}
            />
          </Field>
        </Row>
        <Row>
          <Field>
            <label className="itemLabel">{t('common:amount')}</label>
            <input
              name="amount"
              type="number"
              step="0.01"
              value={tax?.amount}
              onChange={handleTaxChange}
              placeholder={t('common:amount')}
            />
          </Field>
          <Field>
            <label className="itemLabel">{t('form:fields:tax:method')}</label>
            <select
              name="method"
              value={tax?.method}
              onChange={handleTaxChange}
            >
              <option value="default">{t('common:default')}</option>
              <option value="reverse">{t('form:fields:tax:reverse')}</option>
            </select>
          </Field>
        </Row>
      </Part>
    </>
  )
}

export default Tax
