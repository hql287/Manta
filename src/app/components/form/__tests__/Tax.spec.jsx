// Libs
import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';

// Component
import { Tax } from '../Tax.jsx';

// Mocks
const tax = {
  amount: 10,
  tin: '123-456-789',
  mode: 'default',
};
const savedSettings = {
  amount: 10,
  tin: '123-456-789',
  mode: 'default',
};
const t = jest.fn();
const updateFieldData = jest.fn();
const updateSavedSettings = jest.fn();

describe('Note component', () => {
  let wrapper;
  beforeEach(() => {
    wrapper = mount(
      <Tax
        t={t}
        tax={tax}
        savedSettings={savedSettings}
        updateSavedSettings={updateSavedSettings}
        updateFieldData={updateFieldData}
      />
    );
  });

  // PROPS & STATE
  it('receives correct props', () => {
    expect(wrapper.prop('tax')).toEqual(tax);
    expect(wrapper.prop('updateFieldData')).toEqual(updateFieldData);
  });

  // RENDER
  it('renders necessary element', () => {
    expect(wrapper.find('label')).toHaveLength(4);
    expect(wrapper.find('input')).toHaveLength(2);
    expect(wrapper.find('select')).toHaveLength(1);
  });

  // LIFE CYCLE EVENTS
  // TODO
  it('render when necessary', () => {});

  // PRIVATE METHOD
  it('handleInputChange correctly', () => {
    const spy = jest.spyOn(Tax.prototype, 'handleInputChange');
    const wrap = mount(
      <Tax
        t={t}
        tax={tax}
        savedSettings={savedSettings}
        updateSavedSettings={updateSavedSettings}
        updateFieldData={updateFieldData}
      />
    );
    // Changing Amount
    const amount = wrap.find('input').last();
    amount.simulate('change', { target: { value: '20' } });
    expect(spy).toHaveBeenCalled();
    // Changing TIN
    const tin = wrap.find('input').first();
    tin.simulate('change', { target: { value: '987-654-321' } });
    expect(spy).toHaveBeenCalled();
    // Changing Mode
    const mode = wrap.find('select');
    mode.simulate('change', { target: { value: 'reverse' } });
    expect(spy).toHaveBeenCalled();
  });

  // SNAPSHOT
  it('matches snapshot', () => {
    const tree = renderer
      .create(
        <Tax
          t={t}
          tax={tax}
          savedSettings={savedSettings}
          updateSavedSettings={updateSavedSettings}
          updateFieldData={updateFieldData}
        />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
