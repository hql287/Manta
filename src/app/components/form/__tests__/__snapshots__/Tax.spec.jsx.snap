// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Note component matches snapshot 1`] = `
<div
  className="Section__SectionStyle-dgSKSv cwfAAu"
>
  <div
    className="Section__HeaderStyle-chmDx gHBKbw"
  >
    <label
      className="itemLabel"
    />
  </div>
  <div
    className="Part__PartStyle-kAWmVX iwwkMM"
  >
    <div
      className="Part__RowStyle-nLRvo cGeXDT"
    >
      <div
        className="Part__FieldStyle-dkcwja bIqrsv"
      >
        <label
          className="itemLabel"
        />
        <div
          className="Tax__TaxID-jmRTWe QNlQZ"
        >
          <input
            name="tin"
            onChange={[Function]}
            placeholder={undefined}
            type="text"
            value="123-456-789"
          />
        </div>
      </div>
    </div>
    <div
      className="Part__RowStyle-nLRvo cGeXDT"
    >
      <div
        className="Part__FieldStyle-dkcwja bIqrsv"
      >
        <label
          className="itemLabel"
        >
           (%)
        </label>
        <div
          className="Tax__TaxAmount-hmOePo dZzXEm"
        >
          <input
            name="amount"
            onChange={[Function]}
            placeholder={undefined}
            step="0.01"
            type="number"
            value={10}
          />
        </div>
      </div>
      <div
        className="Part__FieldStyle-dkcwja bIqrsv"
      >
        <label
          className="itemLabel"
        />
        <select
          name="method"
          onChange={[Function]}
          value={undefined}
        >
          <option
            value="default"
          />
          <option
            value="reverse"
          />
        </select>
      </div>
    </div>
  </div>
</div>
`;
